### ISSUE TITLE
Better ModelDesc

### ISSUE THREAD

**Issue by ppwwyyxx** [ppwwyyxx/tensorpack#318]
The original design lacks enough consideration and it's not clear how the graph is built, and what one can and cannot do inside build_graph. E.g.:

+ Is it OK to create placeholders inside build_graph?
+ What symbolic functions are allowed to use and what not? (e.g. tf.layers.batch_norm? tf.train.input_producer?)..
+ What to put in `get_inputs` and what not? Is this interface even necessary?
+ __FIXED__ by introducing TowerTrainer, TowerFunc, TowerTensorHandle <del>How to access a tensor a bit later? Because setting `self.xxx` sadly doesn't work (#287), and using the tensor names is not easy. (#315, #317, #442) </del>
+ __RESOLVED__ Use `return cost` for single-cost `ModelDesc`. For other types of models, you need to write your own trainer any way, so you'll build the graph by yourself anyway. <del>On the contrary, `self.cost` needs to be set. This seems very hard-coded, and the reason behind it is that `self.cost` is only set because **some** (but not all) trainers need it. This contract between Model and Trainer needs to be addressed in a clearer way. </del>
+ __FIXED__ <del>What's worse, some examples now actually is using `self.xxx`. Technically they should not rely on this unsupported use.</del>
+ Fancy dynamic stuff might also be hard, but I'm not very familiar.

Some of example use case that is hard or too tricky to do with the current interface:
+ Input data has different layout (needs different placeholder) in training vs inference.
+ Access some tensors in all towers.
+ Mix of data/model parallel. A special case is to create some variables (not reuse) in each tower.

Nothing should be deprecated because the current interface works well for most problems. But I'm thinking about new ones which can expose more of the graph building process to users.

**Comment by vqdang** [comment]
Could you also add a method to just dump out the base graph to Tensorboard? (No training yet, just general structure debugging) I'm using something like below but it requires to comment/uncomment  some part in `batch_norm.py` every time between debugging/training (mostly the `ctx`).

```
def _visualize_graph(self, outdir):
    x = tf.placeholder(tf.uint8, shape=(None, None, None, 3))
    y = tf.placeholder(tf.uint8, shape=(None, None, None, None))
    self._build_graph([x, y])

    with tf.Session() as sess:
        writer = tf.summary.FileWriter(outdir, sess.graph)
        print(sess.run(self.graph_tree))
        writer.close()
```

**Comment by ppwwyyxx** [comment]
Could you be more specific what's wrong with the current graph in tensorboard?
My understanding is that it includes something you don't want to see. But a graph is complicated because neural net training is just so complicated, no way I can automatically guess what you want to see in the graph and what you don't. If you're OK with what your code sample produces, then you can just use it.

What's wrong with batch norm? I think their ops would be grouped under the variable scope so it should look fine.

**Comment by PatWie** [comment]
I really like the current design. It would be frustrating to refactor all models. But it might be helpful if one could write: 

```
Trainer.tower[0].model.var
```
When writing self.var = ...

I do not see a reason for rewriting most stuff. There will always be a cost Tensor.

**Comment by vqdang** [comment]
@ppwwyyxx Sorry for not being clear enough. It's not about all of the ops in the whole graph but the main portion itself. For example:

```
def _build_graph(self, inputs):
    image, label = inputs

    ### Main Network - simple neural network
    with argscope(Conv2D, out_channel=32, kernel_shape=3, nl=tf.nn.relu):
        l = Conv2D('conv0', image, 32, 3, nl=tf.nn.relu)
        l = MaxPooling('pool0', l, 2)
        l = FullyConnected('fc0', l, 512, nl=tf.nn.relu)
        l = Dropout('dropout', l, 0.5)
        l = FullyConnected('fc1', l, 10, nl=tf.identity)

    ### Main Network - complex network with lots of interconnections between layers
    with argscope(Conv2D, out_channel=32, kernel_shape=3, nl=tf.nn.relu):
        l = Conv2D('conv0', image, 32, 3, nl=tf.nn.relu)
        l1, l2, l3 = layer_a('complex_layer_a', l)
        l = layer_b('complex_layer_b', [l, l1, l2])
        l = tf.concat([l, l1, l2, l3], 3)
        l = FullyConnected('fc1', l, 10, nl=tf.identity)

    ### Others stuff in graph for whole training process
    loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits, labels=label)
    loss = tf.reduce_mean(loss, name='xentropy-loss')

    add_moving_summary(loss)
    self.cost = loss
```

And I want to debug the `l` tree, in a simple case, there is no need for visualization before actual training. But in a more complex case, visualization is needed if I want to make sure all the layers (or components) are linked as intended before actually run the trainer. For now, I do this via

```
def _visualize_graph(self, outdir):
    x = tf.placeholder(tf.uint8, shape=(None, None, None, 3))
    y = tf.placeholder(tf.uint8, shape=(None, None, None, None))
    self._build_graph([x, y])

    with tf.Session() as sess:
        writer = tf.summary.FileWriter(outdir, sess.graph)
        print(sess.run(self.graph_tree))
        writer.close() 
```
With `self.graph_tree = l` in `_build_graph`. However, for `_visualize_graph` to run just by
```
model = Model()
model._visual_graph('model_graph')
```
I need to modify whenever variable `ctx` with `ctx = get_current_tower_context()` is used in all the ops in `l` (in my case, it is the `BatchNorm`). Because in case above, there clearly no context of the tower and I'm not sure how to set the context manually. 

In short, some placeholder for similar situation or an official support to visualize part of the whole graph tree. Beside, to actually get to the graph tree in Tensorboard via running the trainer is also taking much longer time and not suit to debug just the main structure I think.

**Comment by ppwwyyxx** [comment]
I still could not see what's the problem with the current graph in tensorboard. In your code you still dumped everything in `_build_graph` to tensorboard. I can see the only difference to what is now done in tensorpack is that 1. you won't dump the backward part 2. you won't dump the inference tower. I would expect these two parts won't make too much mess in tensorboard.

See [docs](http://tensorpack.readthedocs.io/en/latest/tutorial/model.html#use-models-outside-tensorpack) for your problem about the context.

**Comment by PatWie** [comment]
@vqdang Are you looking for (?):

```python
m = Model()
hdrs = m.get_reused_placehdrs()

with TowerContext('', is_training=False):
    m._build_graph(hdrs)
    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())

        # do you stuff here
```

Then the `ctx` is available.

**Comment by ppwwyyxx** [comment]
@PatWie I updated the issue with some more concrete use cases.

**Comment by Skylion007** [comment]
Some other things to consider about how to potentially break and fix ModelDesc:

Partial inference: DiscoGAN is a good example. For prediction you might want to only go from A->B.
Additionally, you may want to the graph to vary significantly in prediction (like running A->B generator multiple times recursively, this is currently rather cumbersome with the predictor).

Loading in a predictor network: Networks that rely on things like Inception score for instance. PyTorch handles this really well by allowing you to just call the model inference of Inception from another model. 

Input Desc: Seems a little redundant to me, how is it different from place holders? Can I only send data to placeholder Tensors? It would be nice to be able to pump data into any part of the subgraph. In which case, what's the point of placeholders at all? Can we inject placeholders into the model on demand? Does this have negative performance implications? Does this violate the principle of least astonishment. 

Easy way to parameterize models. Additionally, well self.XXX is nonstandard, I think it can be extremely useful for building the model. CycleGAN example of using Resblocks come to mind. Defining the depth of the model should be easy to parameter meaning there should be some easy way to send model arguments. Maybe even trainers that can do parameter sweeps? Should trainers be part of the model.

We really should encourage people to use proper namespaces. Should we prefix entire ModelDesc with model names? What if a user want to do this.

I know it's slightly outside the scope of Tensorpack, but can we make ModelDesc composable? Can I have models in my models? Can I have ModelDesc from stuff as simple as Resblocks? Should we have generic ModelDesc for different frameworks?

When and where to call collect_variables can be a little confusing. Should it be baked into the model? Should it not?

Idea for accessing specific Tensors: Encourage Tensornamed namespaces? Force users to use them in the model somehow. Maybe even draft a style guide.

I will say, one of my favorite things about Tensorpack is the ability to learn by example thanks to the very flushed out examples and limited symbolic functions. While I understand maintaining the latter is outside the scope of this project, implementing models can really help find flaws in stuff like trainers and ModelDesc. The example of WGAN comes to mind.

**Comment by PatWie** [comment]
The InputDescr is needed, otherwise the data flow does not know how the data goes into the graph. 
If it wouldn't be there models hard to read as you have to backtrack the cost tensor to the input which sometimes require a large mental stack.

And you can use [Models in Models](https://github.com/ppwwyyxx/tensorpack/blob/master/examples/Saliency/saliency-maps.py#L28) as the constructor is empty. I did this multiple times. Models parameterization is also possible, just add a constructor with the parameter. I used it to build entire recurrent networks and specify the time steps which can varying during interference. 

So it is not as bad as it looks like 😄 

Also the partial inference: you just specify the output tensor and your done?! 
And in TensorFlow you can send data to placeholders only. This is not a limit of tensorpack which even provide a [workaround](https://github.com/ppwwyyxx/tensorpack/blob/master/examples/GAN/InfoGAN-mnist.py#L90).

I just see the issue with the namespace, which depends on which by trainer is used. Maybe a default 'tower0' fixes this.

It might be that I inhaled already too much tensorpack code. Please do not try to make a second ideas or of it and oversimplify things.

**Comment by ppwwyyxx** [comment]
In addition to the above comments to @Skylion007 
Partial inference: yes the graph might vary a lot in prediction and that's one of the use case I put
in this issue.

Loading a network: I don't see any problems with that.

InputDesc: it exists mainly for working with `InputSource`, which is an abstraction of all possible input source,
not only including placeholders, but could switch among tensors, queues, etc.
In tensorflow you cannot pump data into any part of the subgraph when data comes from an op.

Parameterize ModelDesc: as @PatWie said you can do it.

More complicated models: as mentioned there is no plan to add more models.

`collect_variables` is not a function of tensorpack. 
It's an example -- meaning that's something the user (in this case myself), should write.
The user might not write it clear enough but that's unrelated to the design of tensorpack.

**Comment by ppwwyyxx** [comment]
Possible solutions to the problems left:

1. Allow users to build the whole graph by themselves. (i.e. no ModelDesc at all)
This seems to solve all the problems mentioned, and provide best flexibility for complex graphs. 

It's easy to do: currently the trainer does two jobs: setting up the whole graph from ModelDesc (`_setup()`), and run the loop. We just need a trainer which does nothing in `_setup()`.

2. ModelDesc will be only for simple graphs, then. But it's still awkward to access tensors. The ideal way is to simply do `model.towers[0].cost`, or `model.towers['tower0'].cost` for a tensor set by `self.cost = ...`. This is possible but I still have to see how nicely it can be done.

**Comment by ppwwyyxx** [comment]
One general philosophy about design is that: when you pass something to someone that doesn't need it for sure, it's a design problem. Packing a bunch of (possible None) objects together is a way to hide this design problem (because then you only pass one giant object), but the problem still exists.

`ModelDesc` packs three things: `InputDesc`, `build_graph`, `get_optimizer`.
`InputDesc` is only for working with different types of InputSource. Different InputSource creates very different tensors but they share this common metadata. InputDesc is not used for anything else in the whole project.
`get_optimizer` is only used for training, so ideally `ModelDesc` shouldn't be used to inference.

`TrainConfig` packs a number of things. A lot of them are not useful in certain cases, e.g. only multiGPU trainers need towers. A custom trainer may not need model or dataflow.
Packing all stuff together introduces another problem in `TrainConfig`: those stuff cannot depend on each other, although they may logically do. For example, a callback may depend on the outcome of building the model (need to use tensor names).

The new trainer interface in development doesn't depend on either `ModelDesc` or `TrainConfig`. Instead of taking those packaged objects, it takes individual options. `ModelDesc` and `TrainConfig` can then be used through [high-level wrappers](https://github.com/ppwwyyxx/tensorpack/blob/4e6442903b511ae5dd398d8204cd9eebf0729c77/tensorpack/trainv2/interface.py#L44) around the trainers. Then users will get more control over the details when the high-level interface isn't enough, or when a different high-level interface is needed. For example, it will be much easier to train a Keras model with the new trainers.

(nothing is broken for the moment. just sharing some progress)

**Comment by ppwwyyxx** [comment]
With the new trainer API, we can answer the problems at the beginning:
> Is it OK to create placeholders inside build_graph?
What symbolic functions are allowed to use and what not? (e.g. tf.layers.batch_norm? tf.train.input_producer?)..

You can use anything as long as your training iteration can still run. e.g., if you use a placeholder, you need a callback to feed it when the placeholder is a dependency. If you use tensorflow input ops, you may need a callback to start the queues.

> What to put in get_inputs and what not? Is this interface even necessary?

They are used only to setup `InputSource` in `SingleCostTrainer` as well as `InferenceRunner`. You don't have to use `InputDesc` if not working with `InputSource`. But you'd better use `InputSource` for performance reasons.

> How to access a tensor a bit later? Because setting self.xxx sadly doesn't work (#287), and using the tensor names is not easy.

A new abstraction called `TowerTrainer` will cover models whose graph can be built by calling a function (tower function) with input tensors one or more times. `SingleCostTrainer` belongs to `TowerTrainer`. 
The tower function will keep track of the scopes every time it get called, and then the tensors can be accessed. The API is still not ready yet, but should be something like `trainer.tower_func.towers[0]['cost']`, or `trainer.tower_func.towers['tower0']['cost']`.
Of course, you can only access the tenors after the graph has been built. So if using high-level one-shot training wrappers, you will only have the chance to use this interface inside `callback.setup_graph`.

> On the contrary, self.cost needs to be set. This seems very hard-coded, and the reason behind it is that self.cost is only set because some (but not all) trainers need it. This contract between Model and Trainer needs to be addressed in a clearer way.

It is hard-coded. We'll keep it this way. But now `SingleCostTrainer` takes a function which explicitly returns one cost, i.e., the hidden contract become an API requirement. Other trainer doesn't have any notion about "cost".

**Comment by ppwwyyxx** [comment]
Btw. `build_graph(tensorA, tensorB)` looks more natural than `_build_graph(inputs): tensorA, tensorB = inputs`. The latter exists for historical reasons. But maybe it's better to use the first one?

**Comment by PatWie** [comment]
Be careful for these changes, e.g. the Offline-predictor now requires a single tensor as an input or a list. I found this a little bit unintuitive as I now have some cases, where I want to evaluate the perceptual loss only sometimes. So sometimes I feed in two images (anneed a list) and some only one image (without a list). This makes writing the code a little bit cumbersome.

**Comment by ppwwyyxx** [comment]
I think current recommended calling convention for predictors are positional-args, which means you don't need to use list. You should be able to call with `pred(tensorA)` or `pred(tensorA, tensorB)`. Please tell me if this fails. The idea for `build_graph` is similar because I feel this is more intuitive, then `pred([tensorA])` and `pred([tensorA, tensorB])`. But maybe I'm wrong.

**Comment by PatWie** [comment]
Given a List, we can anyway use `func(*my_list)`. So there is no huge difference.

**Comment by ppwwyyxx** [comment]
Sure you can always convert between the two.
Just feel writing `tensorA, tensorB, tensorC = inputs` all the time is redundant.

**Comment by PatWie** [comment]
Update examples to use `build_graph` instead of `_build_graph`?
```python
def build_graph(self, image, label):
```
and 
```python
def _build_graph(self, inputs):
   image, label = inputs
```
are identical.

Currently, the internal API seems to use the nice version of `build_graph(self, image, label)` but the intended way still seems to be `build_graph(self, inputs)`. Shouldn't it be the other way around? The `log_deprecated` is never called in `build_graph`.

**Comment by ppwwyyxx** [comment]
Will happen in the future. Along with that I'm thinking whether it's better to do `return cost` instead of `self.cost = cost`.

**Comment by ppwwyyxx** [comment]
I think the future (and final) changes to `ModelDesc` would be:
1. Prefer `build_graph(self, a, b)` instead of `_build_graph(self, inputs)`
2. Prefer `return cost` instead of `self.cost = cost`
3. Won't automatically apply REGULARIZATION_LOSSES collection if using the new `build_graph+return cost` interface. Keep (?) the old behavior if using the current interface. Reasons explained in [the old thread](https://github.com/ppwwyyxx/tensorpack/pull/81#issuecomment-318299215).

These can probably be smoothly executed without breaking anything.
I found a lot of changes I made recently were to remove some "automatic" parts from the framework (or at least make them more explicit): setting logger directory, automatically load epoch number, clear the inference queue before inference, etc. A lot of these "automatic" behaviors look good at the beginning but end up causing trouble in unexpected scenarios.

**Comment by PatWie** [comment]
Not sure, if `build_graph(self, a, b)` is better than `_build_graph(self, inputs)`. In a recent project, we had something like
```python
def _get_inputs(self):
    InputDescs = []
    for i in range(ONLY_KNOWN_AT_RUNTIME):
        InputDescs.append(InputDesc(tf.float32, (B, H[i], W[i], 3), 'input_%i' % i))
    return InputDescs
```

This works quite nice in the current version, but might not work in the proposed changes. How does `_build_graph` looks like? How do you handle this case:


```python
class Parent(object):
    def __init__(self):
        super(Parent, self).__init__()

    def _get_inputs(self):
        pass

    def _build_graph(self, inputs):
        pass


class Child(Parent):
    def __init__(self, num):
        super(Child, self).__init__()
        self.num = num

    def _get_inputs(self):
        self.inputs = []
        for i in range(self.num):
            self.inputs.append(object)

    def _build_graph(self, a, b):
        print a, b

c = Child(3)
c._get_inputs()
c._build_graph([1, 5, 6])
```

**Comment by ppwwyyxx** [comment]
You can use `*args`, but that's a good point. I'm not confident which one is better.

`return cost` will be used by `ModelDesc` which is intended for single-cost single-optimizer training only.

**Comment by ppwwyyxx** [comment]
Looking further, however, I think letting users to "write a class" is the one thing that makes tensorpack very hard to use at the first sight. Usually you don't use a library by inheritance..

For example, if we could transit to something like:
```python
TrainConfig(
model=ModelFactory(
  inputs_desc=[],
  tower_func=lambda image, label: return cost,
  optimizer=tf.train.GradientDescentOptimizer(0.1)
)
...
)
```
This would probably make everything looks much simpler. The Tensorpack+Keras example is written somewhat like this.

**Comment by PatWie** [comment]
For a second, it sounds like a nice idea to ease the way of using tensor pack for new users.
But again input—desc should be a function. Maybe, this would even allow to directly use tf.placeholders. I think for small examples (Mnist) the function style is much easier to understand. 

But thinking about a bit longer, I feel VERY uncomfortable.
Personally, I like the way of using a class rather than writing multiple lines of a config. Passing tons of lines as a config instead of objects usually smells like a bad design.  In addition, Inputs, Graph, Optimizer are not a loose collection they are strongly connected. In C++ you usually want to pass struct  objects instead of its members to functions, right? Your proposal sounds like dropping c++ and going back to C.

When touching ModelDescr, maybe usingthe design of

```
Class Model
  Def Inputs
    Return tf.placeholders
  Def Graph
    Return cost
  Def Optimizer
    Return Adam
```

 might be a proposal from my side: just rename these methods. Directly using TF.placeholders might be possible as well. You can always place them first in a second graph to get their shapes and just delete this helper graph.

Hopefully, when changing  anything you will still support the class based way. tensorpack is now in a state, where people like me do not want to maintain all old examples -- besides maybe renaming ModelDescr to ModelDescrOld.

**Comment by ppwwyyxx** [comment]
Yes I agree using a class is more powerful, e.g. able to change number of inputs, able to share attributes between these three functions, etc. These concepts are strongly connected and having them in a class makes sense. I thought about how I would write some of the complex models without the class and that doesn't look very good.

In the meantime, having a class makes it hard to write things. Somehow I recently felt tensorpack was too complicated for the easiest tasks. I'll need a "boilerplate.py" to start working -- that's a bit sad. I don't know whether `ModelFactory` can be a solution, but I think both renaming and `tf.placeholder` sound good!

**Comment by ppwwyyxx** [comment]
How about 39fa465655a2767b ? Both methods can be supported together, and I think `inputs()` + `tf.placeholder` look great.

**Comment by PatWie** [comment]
This is awesome and exactly what I meant by the helper-graph :+1: 
Let me now nag about `_build_graph` vs `graph` and `_get_optimizer` vs `optimizer` :-)

**Comment by PatWie** [comment]
Considering `*args` as a default in
```python
class Parent(object):
    def __init__(self):
        super(Parent, self).__init__()

    def inputs(self):
        pass

    def graph(self, *inputs):
        pass


class Child(Parent):
    def __init__(self, num):
        super(Child, self).__init__()
        self.num = num

    def inputs(self):
        self.inputs = []
        for i in range(self.num):
            self.inputs.append(object)

    def graph(self, *inputs):
        image, label = inputs
        print image, label

class Child2(Parent):
    def __init__(self, num):
        super(Child2, self).__init__()
        self.num = num

    def inputs(self):
        self.inputs = []
        for i in range(self.num):
            self.inputs.append(object)

    def graph(self, image, label):
        print image, label

c = Child(3)
c.inputs()
c.graph(1, 5)

c = Child2(3)
c.inputs()
c.graph(1, 5)
```

would make the examples look much more intuitive. Looking forward to something like 

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: mnist-tflayers.py

import os
import argparse
import tensorflow as tf



# Just import everything into current namespace
from tensorpack import *
from tensorpack.tfutils import summary, get_current_tower_context
from tensorpack.dataflow import dataset

IMAGE_SIZE = 28


class Model(ModelDesc):
    def inputs(self):
        """
        Define all the inputs (with type, shape, name) that
        the graph will need.
        """
        return [tf.place_holder(tf.float32, (None, IMAGE_SIZE, IMAGE_SIZE), 'input'),
                tf.place_holder(tf.int32, (None,), 'label')]

    def graph(self, image, label):

        image = tf.expand_dims(image, 3) * 2 - 1

        l = tf.layers.conv2d(image, 32, 3, padding='same', activation=tf.nn.relu, name='conv0')
        l = tf.layers.max_pooling2d(l, 2, 2, padding='valid')
        l = tf.layers.conv2d(l, 32, 3, padding='same', activation=tf.nn.relu, name='conv1')
        l = tf.layers.conv2d(l, 32, 3, padding='same', activation=tf.nn.relu, name='conv2')
        l = tf.layers.max_pooling2d(l, 2, 2, padding='valid')
        l = tf.layers.conv2d(l, 32, 3, padding='same', activation=tf.nn.relu, name='conv3')
        l = tf.layers.flatten(l)
        l = tf.layers.dense(l, 512, activation=tf.nn.relu, name='fc0')
        l = tf.layers.dropout(l, rate=0.5,
                              training=get_current_tower_context().is_training)
        logits = tf.layers.dense(l, 10, activation=tf.identity, name='fc1')

        tf.nn.softmax(logits, name='prob')   # a Bx10 with probabilities

        # a vector of length B with loss of each sample
        cost = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits, labels=label)
        return tf.reduce_mean(cost, name='cross_entropy_loss')  # the average cross-entropy loss

    def optimizer(self):
        return tf.train.AdamOptimizer(0.001)


def get_data():
    train = BatchData(dataset.Mnist('train'), 128)
    test = BatchData(dataset.Mnist('test'), 256, remainder=True)
    return train, test


def get_config():
    dataset_train, dataset_test = get_data()
    # How many iterations you want in each epoch.
    # This is the default value, don't actually need to set it in the config
    steps_per_epoch = dataset_train.size()

    # get the config which contains everything necessary in a training
    return TrainConfig(
        model=Model(),
        dataflow=dataset_train,  # the DataFlow instance for training
        callbacks=[
            ModelSaver(),   # save the model after every epoch
            MaxSaver('validation_accuracy'),  # save the model with highest accuracy (prefix 'validation_')
            InferenceRunner(    # run inference(for validation) after every epoch
                dataset_test,   # the DataFlow instance used for validation
                ScalarStats(['cross_entropy_loss', 'accuracy'])),
        ],
        steps_per_epoch=steps_per_epoch,
        max_epoch=100,
    )


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--gpu', help='comma separated list of GPU(s) to use.')
    parser.add_argument('--load', help='load model')
    args = parser.parse_args()
    if args.gpu:
        os.environ['CUDA_VISIBLE_DEVICES'] = args.gpu

    # automatically setup the directory train_log/mnist-convnet for logging
    logger.auto_set_dir()

    config = get_config()
    if args.load:
        config.session_init = SaverRestore(args.load)
    # SimpleTrainer is slow, this is just a demo.
    # You can use QueueInputTrainer instead
    launch_train_with_config(config, SimpleTrainer())
```

**Comment by ppwwyyxx** [comment]
I'm happy with the current `ModelDesc` interface:

inputs() with placeholders
build_graph(self, a, b) which returns a cost (or nothing for generic `ModelDescBase`)
optimizer()

closing this finally.
